### ISSUE TITLE
Feature Request: Given an S3 bucket with a bucket policy, add/merge the k9 policy with this policy

### ISSUE THREAD

**Issue by brentryan** [k9securityio/k9-cdk#12]
I've found that some of the AWS CDK constructs modify the S3 Bucket policy automatically during creation. This is problematic if you overwrite the policy with the new k9 policy because some of the CDK constructs generate custom resources that take actions against the bucket which will fail to work if they're not included in the final policy.

For example:
```
const bucket = new Bucket(this, 'Bucket', {
  removalPolicy: RemovalPolicy.DESTROY,
  autoDeleteObjects: true,
});
```

This will add a bucket policy to enable autoDelete to work.

I think we need a k9 api that takes the bucket.policy and then adds to this policy instead of always creating a new policy.

**Comment by skuenzli** [comment]
Thank you for filing this issue Brent.

We'll take a look at the auto delete objects use case and see how we can make it work.

We'll also take a look at how we might be able to mutate an existing policy.  The main reasons k9 generates the policy from scratch are to:
* be able to reason about the full policy and write the `DenyEveryoneElse` statement correctly
* enable composition into existing code, i.e. don't want to force people to adopt a 3rd-party Bucket construct

**Comment by skuenzli** [comment]
The S3Bucket construct uses the [`enableAutoDeleteObjects` method](https://github.com/aws/aws-cdk/blob/v1.91.0/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1800) to add statements to the resource policy:
```
  private enableAutoDeleteObjects() {
    const provider = CustomResourceProvider.getOrCreateProvider(this, AUTO_DELETE_OBJECTS_RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'auto-delete-objects-handler'),
      runtime: CustomResourceProviderRuntime.NODEJS_12,
      description: `Lambda function for auto-deleting objects in ${this.bucketName} S3 bucket.`,
    });

    // Use a bucket policy to allow the custom resource to delete
    // objects in the bucket
    this.addToResourcePolicy(new iam.PolicyStatement({
      actions: [
        ...perms.BUCKET_READ_ACTIONS, // list objects
        ...perms.BUCKET_DELETE_ACTIONS, // and then delete them
      ],
      resources: [
        this.bucketArn,
        this.arnForObjects('*'),
      ],
      principals: [new iam.ArnPrincipal(provider.roleArn)],
    }));
```
This code creates a custom CFn resource which runs a Lambda function that lists and deletes objects in the bucket during removal.

S3Bucket's `addToResourcePolicy` is called in a number of places.  In general, it would work to make a k9 bucket policy then iterate through document and add each of the statements.  The tricky bit is the `DenyEveryoneElse`.

The AutoDeleteObjects custom resource creates its own Role to execute the function.  And that role name won't be known until after the stack is created.  Now, we know the naming convention of the role, so we could carve out an exclusion for it, and possibly other roles named after the stack.  Will investigate some more and propose a solution.

**Comment by skuenzli** [comment]
Thanks for the additional information.  Only allowing auto-delete in non-prod makes sense.

I have prototyped a solution that merges the k9 policy statements into an existing policy if it already exists.  That's more difficult than it sounds because AWS CDK offers a nearly write-only API on top of private state, so it's difficult to inspect what's already there without transiting policies & statements through `toJSON()`.

**Comment by skuenzli** [comment]
@brentryan - would love for you to take a look at #13 if you have a chance.

**Comment by skuenzli** [comment]
Addressed this feature request in #13 and released v0.7.0
